File: ledger/src/main/scala/com/wonderland/ledger/repository/AccountRepository.scala
package com.wonderland.ledger.repository

import cats.syntax.all.*
import com.wonderland.ledger.domain.*
import doobie.*
import doobie.implicits.*
import doobie.postgres.implicits.*
import java.util.UUID

class AccountRepository:
  def create(cmd: CreateAccount): ConnectionIO[Account] =
    val id = UUID.randomUUID()
    sql"""
      INSERT INTO accounts (id, name, balance, currency)
      VALUES ($id, ${cmd.name}, 0, ${cmd.currency})
      RETURNING id, name, balance, currency, created_at
    """.query[Account].unique

  def findById(id: UUID): ConnectionIO[Option[Account]] =
    sql"SELECT id, name, balance, currency, created_at FROM accounts WHERE id = $id"
      .query[Account].option

  def findByIdForUpdate(id: UUID): ConnectionIO[Option[Account]] =
    sql"SELECT id, name, balance, currency, created_at FROM accounts WHERE id = $id FOR UPDATE"
      .query[Account].option

  def findAll: ConnectionIO[List[Account]] =
    sql"SELECT id, name, balance, currency, created_at FROM accounts ORDER BY created_at DESC"
      .query[Account].to[List]

  def updateBalance(id: UUID, newBalance: BigDecimal): ConnectionIO[Unit] =
    sql"UPDATE accounts SET balance = $newBalance WHERE id = $id"
      .update.run.void

File: hesab-ketab/src/main/java/io/github/bmd007/wonderland/hesab_ketab/repository/AccountRepository.java
package io.github.bmd007.wonderland.hesab_ketab.repository;

import io.github.bmd007.wonderland.hesab_ketab.domain.Account;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
@RequiredArgsConstructor
public class AccountRepository {

    private final JdbcClient jdbc;

    public void save(Account snapshot) {
        jdbc.sql("""
                INSERT INTO accounts (id, name, balance, version)
                VALUES (:id, :name, :balance, :version)
                ON CONFLICT (id) DO UPDATE
                SET balance = :balance, version = :version
                """)
            .param("id", snapshot.id())
            .param("name", snapshot.name())
            .param("balance", snapshot.balance())
            .param("version", snapshot.version())
            .update();
    }

    public Optional<Account> findById(UUID id) {
        return jdbc.sql("SELECT id, name, balance, version, created_at FROM accounts WHERE id = :id")
            .param("id", id)
            .query(Account.class)
            .optional();
    }

    public Optional<Account> findByIdForUpdate(UUID id) {
        return jdbc.sql("SELECT id, name, balance, version, created_at FROM accounts WHERE id = :id FOR UPDATE")
            .param("id", id)
            .query(Account.class)
            .optional();
    }

    public List<Account> findAll() {
        return jdbc.sql("SELECT id, name, balance, version, created_at FROM accounts ORDER BY created_at DESC")
            .query(Account.class)
            .list();
    }
}
